---
title: "REPORT"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this part, we will do quality control and some normalizatation work base on TPM data.

## Three statistics to identify outliers

### RLE

RLE(Relative Log Expression)is assume that most expressions in a sample should be near a mean value, only a few genes have differential expression, which means higher or lower than other genes. To calculate it, for each gene *i* , calculate it median in *N* sample as *Medi* , for each sample *j* , and expression value *eij*, count the difference between *eij* and *Medi* : *deij = eij-Medi* , than boxplot each sample base on *deij* and sort by IQR(interquartile range), the sample with lager IQR is more likely to be an outlier. 

### Hierarchical clustering

We use(1 - spearman's correlation) as distances between samples and assume samples should be homogeneous, so all samples should have short distances between others and clustered homogeneously. So samples far away others may be outliers. So we first use distances which is 1-spearmen correlation to hierarchically clustered, than use top 100 gene sort by variance to calculate Mahalanobis Distance, then a chi2 p-value will be calculated based on mahalanobis distance. Then clusters with $ \geq 60\%$ samples with Bonferroni-corrected p-values < 0.05 were marked as outliers.

### D-statistic

For each sample, D-statistic represent the average correlation between its expression and other samples. A sample with lower D-statistic is more likely to be an outlier.



## Raw data's feature

```{r init and print raw data, echo=FALSE, warning=FALSE, results='hide', message=FALSE}
#TODO read config file
library(rjson)
if (!require("RColorBrewer")){
  install.packages("RColorBrewer")
  library(RColorBrewer)
}
json_data<-fromJSON(file="Tool.config")

appendInfo <- toupper(json_data$config$transcriptome$transcriptomeQC$covariants)=="TRUE"
appendInfoFile <- json_data$config$transcriptome$transcriptomeQC$covariants_file
csvfile <- "TPM_matrix.csv"
RLEFilterPercent <-json_data$config$transcriptome$transcriptomeQC$RLEFilterPercent
DSFilterPercent <- json_data$config$transcriptome$transcriptomeQC$DSFilterPercent
pvalues.cut <- json_data$config$transcriptome$transcriptomeQC$hcluster$pvalues.cut
treesNum <- json_data$config$transcriptome$transcriptomeQC$hcluster$cluster_level


library(ape)
library(reshape2)
library(tidyverse)
TPM_data <- read.csv(csvfile,header = TRUE,row.names = 1)
keep_genes_idx <- (rowMeans(TPM_data>0.1)>0.2) 
TPM_data = TPM_data[keep_genes_idx,]

if(appendInfo)
{
  clinical <- read.table(appendInfoFile,header = TRUE,stringsAsFactors = TRUE)
  rownames(clinical) <- clinical$mrna_id
  idx <- intersect(colnames(TPM_data),rownames(clinical))
  clinical <- clinical[idx,]
  TPM_data <- TPM_data[,idx]
}

rawdata <- TPM_data
RLEFilterLength <- RLEFilterPercent*ncol(TPM_data)
DSFilter <- DSFilterPercent*ncol(TPM_data)


#print raw data's statistic including RLE,D and hclust
#RLE
logtpm = log10(TPM_data + 1e-4)
rle=logtpm-apply(logtpm, 1, median) # change "/" to "-" so that we got log(fold-change) which centered on 0 on the RLE plot.
iqr = apply(rle,2,IQR)
rle=melt(cbind(ID=rownames(rle), rle), variable.name = "Sample",value.name ="TPM", id="ID")
names(rle) <- c("feature","Sample","TPM")
bymedian <- with(rle, reorder(Sample, TPM, IQR))  # sort by IQR
par(mar=c(3,3,3,3))
boxplot(TPM ~ bymedian, data=rle, outline=F, las=2, boxwex=1, col='gray', cex.axis=0.3, main="Raw data Relative Log Expression", xlab="", ylab="RLE", frame=F)
ExpPerSample <- nrow(TPM_data)
RLEFilterList <- unique(bymedian[((length(bymedian)-(ExpPerSample*RLEFilterLength))+1):length(bymedian)]) #filtered

#hcluster
sampleDists = 1 - cor(TPM_data, method='spearman')
hc=hclust(as.dist(sampleDists),method = "complete")
hcphy = as.phylo(hc)
if(appendInfo && is.element("mrna_id",colnames(clinical)) && is.element("batch",colnames(clinical))){
 co = clinical$batch[match(hcphy$tip.label, clinical$mrna_id)]
 co = as.factor(co)
 co_length = length(levels(co))
 co = brewer.pal(co_length,"Set3")
 par(mar=c(3,3,3,3))
 plot(hcphy, tip.col = co, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Raw Data clusting(Spearman - Cor.)")
 #Xtext = c("batch 0", "batch 1","batch 2","batch 3","batch 4","batch 5","batch 6","batch 7","batch 8")
 Xtext = paste("batch",1:co_length)
 legend('bottomleft',pch=21,Xtext, col='white',pt.bg=co, cex=.5)
} else {
  plot(hcphy, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Raw Data clusting(Spearman - Cor.)")
}



logtpm = log10(TPM_data + 1) 
ntop <- 100
Pvars <- apply(logtpm, 1, var)
select <- order(Pvars, decreasing =TRUE)[seq_len(min(ntop, length(Pvars)))]
MD_matrix <- logtpm[select, ]
MahalanobisDistance = mahalanobis(t(MD_matrix), colMeans(t(MD_matrix)), cov(t(MD_matrix))) 
# Note: t(MD_matrix)=sample_row*gene_column, Manhalanobis() returns one vector with length=row number
pvalues = pchisq(MahalanobisDistance, df=nrow(MD_matrix), lower.tail=F)
pvalues.adjust = p.adjust(pvalues, method ="bonferroni") # adjusted pvalues for each sample
pvalues.low <- pvalues.adjust[pvalues.adjust<pvalues.cut]

HCoutliers <- character()
for(x in c(1:treesNum)){
  trees <- cutree(hc,k=x)
  idx <- c(1:x)#which tree is checking
  for(i in idx)
  {
    group <- hc$labels[which(trees == i)]
    if(sum(group %in% names(pvalues.low))/length(group) >= 0.6)
    {
      HCoutliers <- union(HCoutliers,group)
    }
  }
}



#D-s
D = apply(1-sampleDists, 1, median)
hist(D, breaks=100, ylab="Number of samples", xlab="D-statistic", main="Raw Data Histogram of D-statistic")
DSFilter <- sort(D)[DSFilter]
D<-as.data.frame(D)
D<-data.frame(Sample = rownames(D),D = D$D)
D_filterList = D%>%filter(D <= DSFilter)
D_filterList <- D_filterList$Sample
```

### Outliers voted by RLE
```{r , echo=FALSE, warning=FALSE, message=FALSE}
RLEFilterList<-as.character(RLEFilterList)
RLEFilterList
```

### Outliers voted by Hcluster
```{r , echo=FALSE, warning=FALSE, message=FALSE}
HCoutliers
co1 = hc$labels%in%HCoutliers
co1[which(co1 == "FALSE")]<-"gray0"
co1[which(co1 == "TRUE")]<-"red"
par(mar=c(3,3,3,3))
plot(hcphy, tip.col = co1, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Mark Outlier on cluster graphic")
Xcol = c("gray0", "red")
Xtext = c("Normal Sample", "Outliers")
legend('bottomleft',pch=21,Xtext, col='white',pt.bg=Xcol, cex=1)
```



### Outliers voted by D-statistic
```{r , echo=FALSE, warning=FALSE, message=FALSE}
D_filterList<-as.character(D_filterList)
D_filterList
```

### Their union is picked as outliers
```{r , echo=FALSE, warning=FALSE, message=FALSE}
outliersList <- intersect(RLEFilterList,intersect(HCoutliers,D_filterList))
outliersList
```


```{r gender mismatch, echo=FALSE, warning=FALSE, results='hide', message=FALSE}
if(appendInfo && is.element("sex",colnames(clinical))){
  #reserved words:msex,mrna_id,mrna_id_batch,age_death,pmi

    # Gender mismatch
    library(e1071)
    mismatchFilterList = list()
    message("generating gender-match plot...")
    chrX='ENSG00000229807'  # XIST
    chrY='ENSG00000129824'  # RPS4Y1
    d=as.data.frame(t(logtpm[c(chrX,chrY),])); 
    colnames(d) <- c("chrX","chrY")
    d$Sex = clinical$sex[match(rownames(d), clinical$mrna_id)]
    plot(d$chrX, d$chrY, xlab="Expression of XIST", ylab="Expression of RPS4Y1", col= 'white',bg=ifelse(d$Sex==0,'red','blue'), pch=21, bty="n", main="Gender-specific expression")
    legend('bottomleft',pch=21,c("Female","Male"), col='white',pt.bg=c("red","blue"), bty='n', cex=.5)
    model1 = svm(Sex ~ .,data = d,type = 'C-classification')
    plot(model1,d)
    pSex = predict(model1)
    pSex = as.numeric(pSex)
    pSex = pSex - 1
    matchlist = (pSex == d$Sex)
    mismatchList = which(matchlist==FALSE)
    clusterFilterList = rownames(d)[mismatchList]
    clusterFilterList <- as.character(clusterFilterList)
    outliersList<- union(outliersList,clusterFilterList)
    
  
}
```


```{r if no appendInfo, echo=FALSE, warning=FALSE, results='hide', message=FALSE}
if(!appendInfo){
  print("with no append information, the pipeline is ended")
  outliersIndex <- which(colnames(TPM_data) %in% outliersList)
  if(!length(outliersIndex) == 0){
    TPM_data <- TPM_data[,-outliersIndex]
  }
  keep_genes_idx <- (rowSums(TPM_data>0.1)>=10) 
  expr= TPM_data[keep_genes_idx,]
  expr=log10(expr+1e-4)  # so row value of 0 will be -4 in the transformed value
  # outlier correction: quantile normalization with order preserved. Now TPM is changed to rank normalized gene expression.
  m=apply(expr, 1, mean); 
  sd=apply(expr, 1, sd)
  expr = t(apply(expr, 1, rank, ties.method = "average"));
  #expr = qnorm(expr / (ncol(expr)+1));  # to standard normalization
  expr = qnorm(expr / (ncol(expr)+1), mean=m, sd=sd)  # or, to preserve the mean and sd of each gene
  rm(m,sd)
  expr = as.data.frame(expr)
  
  write.table(format(expr, digits=4,nsmall=4), file = "expression.postQC.xls", sep="\t", col.names = NA, quote=F,row.names = TRUE)
  
}

```


```{r adjust covariates and combat and sva, echo=FALSE, warning=FALSE, results='hide', message=FALSE}
if(appendInfo)
{
outliersIndex <- which(colnames(TPM_data) %in% outliersList)
if(!length(outliersIndex) == 0){
  TPM_data <- TPM_data[,-outliersIndex]
}
if(is.element("mrna_id",colnames(clinical)) || is.element("batch",colnames(clinical))){
  keep_genes_idx <- (rowSums(TPM_data>0.1)>=10) 
  expr= TPM_data[keep_genes_idx,]
  expr=log10(expr+1e-4)  # so row value of 0 will be -4 in the transformed value
  # outlier correction: quantile normalization with order preserved. Now TPM is changed to rank normalized gene expression.
  m=apply(expr, 1, mean); 
  sd=apply(expr, 1, sd)
  expr = t(apply(expr, 1, rank, ties.method = "average"));
  #expr = qnorm(expr / (ncol(expr)+1));  # to standard normalization
  expr = qnorm(expr / (ncol(expr)+1), mean=m, sd=sd)  # or, to preserve the mean and sd of each gene
  rm(m,sd)
  
  expr = as.data.frame(expr)
  covs = clinical
  rownames(covs) = covs$mrna_id
  covs = covs[names(expr),]
  
  covs$batch = as.factor(covs$batch)

  if(is.element("sex",colnames(covs)))
  {
  	covs$sex = as.factor(covs$sex)
  }
  if(is.element("age_death",colnames(covs)))
  {
    covs$age_death = as.numeric(covs$age_death)
  }
  library(sva)
  message("# adjusting expression with covariates...")
  ######################
  #Combat remove batch effects
  modcombat = model.matrix(~1, data = covs)
  combat_expr = ComBat(dat=as.matrix(expr), batch=covs$batch, mod = modcombat)
  par(mar=c(3,3,3,3))
  sampleDists_combat = 1 - cor(combat_expr, method='spearman')
  hc_combat = hclust(as.dist(sampleDists_combat),method = "complete")
  hcphy_combat = as.phylo(hc_combat)
  plot(hcphy_combat, tip.col = co, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Hcluster after combat")
  #Xcol = c("gray0", "red", "orange", "yellow", "green", "blue", "purple", "wheat", "violet")
  Xtext = paste("batch",1:co_length)
  legend('bottomleft',pch=21,Xtext, col='white',pt.bg=co, cex=.5)
  
  #####################
  #sva adjusted
  Mod = model.matrix(~.-batch-mrna_id, data=covs) # full model(adjustment variables + variables of interest)
  #Mod = model.matrix(~batch, data=covs)
  Mod0 = model.matrix(~1,data=covs) # Null model (vairables of interest)

  svaobj = sva(as.matrix(combat_expr),Mod, Mod0)
  fsvaobj = fsva(dbdat=as.matrix(combat_expr),mod=Mod,sv=svaobj,   newdat=as.matrix(combat_expr))
  residuals = fsvaobj$db

  message("# run RLE on SVA normalized quantification data ...")
  #####################
  #residuals <- combat_expr
  ## RLE after SVA
  res=data.frame(residuals, check.names = F)
  rle=res-apply(res, 1, median) # after SVA
  rle=melt(cbind(ID=rownames(rle), rle), variable.name = "Sample",value.name ="TPM", id="ID")
  bymedian <- with(rle, reorder(Sample, TPM, IQR))  # sort by IQR
  par(mar=c(7,3,3,1))
  boxplot(TPM ~ bymedian, data=rle, ylim=c(-4,4), outline=F, las=2, boxwex=1, col='gray', cex.axis=0.5, main="Final Data RLE", xlab="", ylab="Relative log expression (RLE)")
  abline(h=0, col='red',lwd=1)
  
  ## mean gene expression distribution before and after SVA
  par(mfrow=c(2,1))
  hist(apply(expr,1,mean), breaks=100, xlab="Rank normalized expression log10(TPM)", main="Expression distribution before adjustment")
  hist(apply(residuals,1,mean), breaks=100, xlab="Rank normalized expression log10(TPM)", main="Expression distribution after adjustment")
  ## clustering after sva
  sampleDists_sva = 1 - cor(residuals, method='spearman')
  hc_sva = hclust(as.dist(sampleDists_sva),method = "complete")
  hcphy_sva = as.phylo(hc_sva)
  
  par(mfrow=c(1,1))
  par(mar=c(3,3,3,3))
  plot(hcphy_sva, tip.col = co, type = "unrooted", cex=.2, lab4ut='axial',underscore = T, main="Final data clusting")
  #Xcol = c("gray0", "red", "orange", "yellow", "green", "blue", "purple", "wheat", "violet")
  #Xtext = c("batch 0", "batch 1","batch 2","batch 3","batch 4","batch 5","batch 6","batch 7","batch 8")
  Xtext = paste("batch",1:co_length)
  legend('bottomleft',pch=21,Xtext, col='white',pt.bg=co, cex=.5)
  }
  message("# save final quantification data into file")
  ######################
  write.table(format(residuals, digits=4,nsmall=4), file = "expression.postQC.xls", sep="\t", col.names = NA, quote=F,row.names = TRUE)
  rownames(clinical) = clinical$mrna_id
  clinical_RNAseqSamples = clinical[colnames(residuals),]
  write.table(clinical_RNAseqSamples, file = "clinical_RNAseqQCedSamples.txt", sep="\t", col.names=TRUE, quote=F, row.names=FALSE)
}

```
